File: 44-wildcard-matching/wildcard-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        n,m=len(s),len(p)
        dp=[[-1]*(m+1) for _ in range(n+1)]
        # def rec(i,j):
        #     if i==n and j==m:
        #         return True
        #     if j==m:
        #         return False
        #     if i==n:
        #         for k in range(j,m):
        #             if p[k]!='*':
        #                 return False
        #         return True
        #     if dp[i][j]!=-1:
        #         return dp[i][j]
        #     if p[j]=='?':
        #         val=rec(i+1,j+1)
        #         dp[i][j]=val
        #         return dp[i][j]
        #     if p[j]=='*':
        #         dp[i][j]=rec(i,j+1) or rec(i+1,j)
        #         return dp[i][j]
        #     if s[i]==p[j]:
        #         dp[i][j]=rec(i+1,j+1)
        #         return dp[i][j]
        #     return False
        found =True
        for j  in range(1,m+1):
            if p[j-1]!='*':
                found=False
            dp[0][j]=found
        for i in range(n+1):
            dp[i][0]=False
        dp[0][0]=True
        
        for i in range(1,n+1):
            for j in range(1,m+1):
                if p[j-1]=='?':
                    dp[i][j]=dp[i-1][j-1]
                elif p[j-1]=='*':
                    dp[i][j]=dp[i-1][j] or dp[i][j-1]
                elif s[i-1]==p[j-1]:
                    dp[i][j]=dp[i-1][j-1]
                else:
                    dp[i][j]=False


        return dp[n][m]

        

File: 44-wildcard-matching/wildcard-matching.cpp
bool rec(int i,int j,string &s,string &p,vector<vector<int>>&dp){
    if(i>=s.size()){
        if(j>=p.size())return true;
        for(int k=j;k<p.size();k++){
            if(p[k]!='*')return false;
        }
        return true;
    }
    if(j>=p.size() && i<s.size())return false;

    if(dp[i][j]!=-1) return dp[i][j];
    
    if(p[j]=='?'){
        return dp[i][j]=rec(i+1,j+1,s,p,dp);
    }
    else if(p[j]=='*'){
        
        bool empty_space= rec(i,j+1,s,p,dp);
        bool single_char=rec(i+1,j+1,s,p,dp);
        bool more_than_one=rec(i+1,j,s,p,dp);
        return dp[i][j]=empty_space or single_char or more_than_one; 
    }
    else if(s[i]==p[j]){
        return dp[i][j]=rec(i+1,j+1,s,p,dp);
    }
    else{
        return dp[i][j]=false;
    }
}

class Solution {
public:
    bool isMatch(string s, string p) {

        int n=s.size();
        int m=p.size();
        vector<vector<int>>dp(n,vector<int>(m,-1));

        return rec(0,0,s,p,dp);
        
    }
};